File: 004.숫자고르기/choose_number.kt
import kotlin.system.exitProcess

fun dfs(
  sIdx: Int,
  cIdx: Int,
  iArr: Array<Int>,
  vArr: Array<Boolean>
): Boolean {
  //println("sIdx: $sIdx, cIdx: $cIdx")
  val value: Int = iArr[cIdx]
  if (vArr[value] == false) {
    vArr[cIdx] = true
    val result: Boolean = dfs(sIdx, value, iArr, vArr)
    return result
  } else {
    return value == sIdx
  }
}

fun main() {
  val num: Int = readLine()?.toInt() ?: exitProcess(0)
  val inputArr = Array(num+1) {0}
  var visitedArr = Array(num+1) {false}
  var result: Boolean = false
  var resultCount: Int = 0
  var resultArr: ArrayList<Int> = ArrayList()
  
  for(i: Int in 1..num) {
    inputArr[i] = readLine()?.toInt() ?: exitProcess(0)
  }
  
  for(i: Int in 1..num) {
    //println("--------------------")
    visitedArr = Array(num+1) {false}
    result = dfs(i, i, inputArr, visitedArr)
    
    if (result) {
      resultCount++
      resultArr.add(i)
    }
  }
  
  println(resultCount)
  for(i: Int in resultArr.indices) {
    println(resultArr[i])
  }
  
}

File: 004.숫자고르기/choose_number.py

def dfs(s_idx, c_idx, iList, vList):
  #print("c_idx: " + str(c_idx))
  #print(vList)
  # 다음갈곳
  value = iList[c_idx]

  if (vList[value] == False):
    # 방문표시
    vList[c_idx] = True
    # 방문안했으면 GOGO
    result = dfs(s_idx, value, iList, vList)
    return result
  else:
    if (value == s_idx):
      return True
    else:
      return False

num = int(input())
inputList = [0]

for n in range(1, num+1):
  inputList.append(int(input()))

resultList = []
resultCount = 0

for i in range(1, num+1):
  xNum = i
  yNum = inputList[i]
  
  visitedList = [False] * (num+1)
  #print("---------------------")
  result = dfs(i, i, inputList, visitedList)
  #print(result)

  if result:
    resultCount += 1
    resultList.append(i)

resultList.sort()

print(resultCount)
for item in resultList:
  print(item)